MINOR: clean up node and store sensors

vvcephei opened the pull request:
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vvcephei committed:
MINOR: tracking for node and store metrics

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java:
```diff
@@ -107,18 +84,23 @@ public void addChild(final ProcessorNode<?, ?> child) {
+            if (processor != null) {
+                processor.init(context);
+            }
+            nodeMetrics.nodeCreationSensor.record(time.nanoseconds() - startNs);
```
These changes are a bit tangential, but while looking at the metrics, I realized we were jumping through quite a few hoops to do something simple, so I simplified these latency measurements.

mjsax replied:
I kinda see you point, however, I am wondering if there would be any runtime difference? Should the compiler be smart enough to inline the code and avoid the method calls?

vvcephei replied:
I'm not sure; it's not just a method call, but also a call to an anonymous Runnable instance we created. And one of those referenced (key/value) fields on ProcessorNode (which we put there just to pass to the Runnable). 

The runtime performance is certainly not slower, but I'd have to profile it to see if it's faster. I'm interested in doing this not so much for the performance gain (which I'd hope is minimal) as for the readability/simplicity, which is significantly improved.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java:
```diff
@@ -107,18 +84,23 @@ public void addChild(final ProcessorNode<?, ?> child) {
-            runAndMeasureLatency(time, closeDelegate, nodeMetrics.nodeDestructionSensor);
+            if (processor != null) {
+                processor.close();
+            }
```
By my reckoning, we are trying to record the "destroy" time as a latency metric, but line 104 unloads the sensor and all the related metrics immediately after recording the destroy latency.

I think this means that the destroy metric has never actually been observable (unless you're super lucky to sample it concurrently between the record and remove.

In that case, we apparently don't need it, so I removed it. Alternatively, we can keep it and not call `nodeMetrics.removeAllSensors()`, or remove all but the destroy sensor...

Or we can leave it as it was, but having nonsense code gives me a funny feeling even when it's harmless.

Of course, please let me know if I've simply mis-understood what was happening here.

mjsax replied:
Interesting find. However I am wondering if we can simply remove a metric? It's a public API change and should require a KIP including deprecation period...

vvcephei replied:
Yeah, I'm not sure either. On one hand, it's "dead code" in that it's never updated until it's removed. But on the other hand, it's created when the node is, so it's actually present for all nodes (just always reporting 0).

I can leave it in, since it's harmless, but I wanted to have the discussion, since I'm 99% sure it's also harmless to take it out.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java:
```diff
@@ -119,15 +121,43 @@ public final void removeAllTaskLevelSensors(final String taskName) {
                                          final Sensor.RecordingLevel recordingLevel,
                                          final Sensor... parents) {
-        final String key = threadName + "." + taskName + "." + cacheName;
+        final String key = threadName + "." + taskName + ".cache." + cacheName;
```
now that there are other sensors with the same path length, it seems safer to namespace them all.

Changing the sensor name shouldn't impact anything.

mjsax replied:
this is the name pattern as for `.node.` -- should we have  method that computes the name passing in `node` or `cache` ?

vvcephei replied:
¯\\_(ツ)_/¯ 

It's really trivial logic, so I'm not sure of the value of adding an extra layer of indirection.

mjsax replied:
It's not about being trivial or not, but about consistency. Code duplication has the danger to only update one copy of the code...

vvcephei replied:
I opted for direct string concatenation here, since it's flatter, and therefore easier to reason about, and in this case, I thought that a separate method would basically offer the same pros and cons as concatenation.

That's also why I flattened the other sensor names, the method calls (two of them) were obscuring the fact that one of the portions was totally unused. Once it's flat, you can see directly what string is produced.

Also contributing factors: we need to append one more part to the key after the fact to get the "fullSensorName", and each of these are completely internal and need to be declared in exactly two places, the xLevelSensor and the removeAllXLevelSensors methods.

It seems like concatenation is slightly better, but I'm open to changing it if you feel strongly about it.

vvcephei replied:
I've reconsidered this stance, and added methods to create the "prefix" key for each level of sensor.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java:
```diff
@@ -152,6 +182,41 @@ public final void removeAllCacheLevelSensors(final String taskName, final String
+        }
+    }
+
+    // TODO: should we use this method when the state store gets closed?
```
should we? (we don't currently unload any sensors when we close a store, although all the children of a task-level sensor get unloaded with the task)

guozhangwang replied:
We should, and I think it was actually a bug that we did not do that before.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java:
```diff
@@ -230,38 +288,23 @@ public Sensor addLatencyAndThroughputSensor(final String scopeName,
-                                                final String operationName,
-                                                final Sensor.RecordingLevel recordingLevel,
-                                                final String... tags) {
-
```
The task-scoped `addLatencyAndThroughputSensor` is unused now, so I've removed it.

FWIW, the non-task-scoped `addLatencyAndThroughputSensor` is unused in our code base, but it's part of the public `StreamsMetrics` interface, so I've left it and its tests. Perhaps we can have a KIP some time to remove it.

guozhangwang replied:
`addLatencyAndThroughputSensor` and others in `StreamsMetrics` are for users to register their customized metrics, so we cannot remove it..

guozhangwang replied:
Note that for now these customized metrics will only be add on the "thread-level" though the tag map contains two: the scopeName = entityName, and threadName = [thread-id].

vvcephei replied:
I didn't follow that last comment. These public sensor methods create both a thread-level "operationName" sensor and a "entityName"-level "operationName" sensor.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java:
```diff
@@ -77,79 +81,27 @@
+        rangeTime = createTaskAndStoreLatencyAndThroughputSensors(DEBUG, "range", metrics, metricsGroup, taskName, name(), taskTags, storeTags);
+        flushTime = createTaskAndStoreLatencyAndThroughputSensors(DEBUG, "flush", metrics, metricsGroup, taskName, name(), taskTags, storeTags);
+        deleteTime = createTaskAndStoreLatencyAndThroughputSensors(DEBUG, "delete", metrics, metricsGroup, taskName, name(), taskTags, storeTags);
+        final Sensor restoreTime = createTaskAndStoreLatencyAndThroughputSensors(DEBUG, "restore", metrics, metricsGroup, taskName, name(), taskTags, storeTags);
```
These lines are a little long, but they're also very uniform. Should I split them?

mjsax replied:
I am fine both ways.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/Sensors.java:
```diff
@@ -0,0 +1,46 @@
+import static org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.addAvgMaxLatency;
+import static org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.addInvocationRateAndCount;
+
+public class Sensors {
```
Just creating a place where all the state stores can access this method without exposing it to the whole code base.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/Sensors.java:
```diff
@@ -0,0 +1,48 @@
+import static org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.addAvgMaxLatency;
+import static org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.addInvocationRateAndCount;
+
+public final class Sensors {
```
Just creating a place where all the state stores can access this method without exposing it to the whole code base.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java:
```diff
@@ -103,20 +103,14 @@ public void testMetrics() {
         node.init(context);
 
-        final String[] latencyOperations = {"process", "punctuate", "create", "destroy"};
+        final String[] latencyOperations = {"process", "punctuate", "create"};
```
See above for the question about the "destroy" sensor.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java:
```diff
@@ -103,20 +103,14 @@ public void testMetrics() {
-            assertNotNull(metrics.getSensor("name-mock.0_0." + operation));
-        }
-        assertNotNull(metrics.getSensor("name-mock.0_0." + throughputOperation));
-
```
We don't need to verify the sensors themselves, since we verify the metrics (which are the actual interface)

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsMetricsImplTest.java:
```diff
@@ -52,10 +51,10 @@ public void testRemoveSensor() {
         streamsMetrics.removeSensor(sensor1a);
 
-        final Sensor sensor2 = streamsMetrics.addLatencyAndThroughputSensor(taskName, scope, entity, operation, Sensor.RecordingLevel.DEBUG);
+        final Sensor sensor2 = streamsMetrics.addLatencyAndThroughputSensor(scope, entity, operation, Sensor.RecordingLevel.DEBUG);
```
Testing the public interface instead of the (unused and now removed) internal one.

vvcephei commented:
@guozhangwang @mjsax 

During my last iteration on the metrics, I stopped short of updating the node and state sensors to use the new sensor tracking mechanism.

I need them now for the proposed metrics in KIP-328, so I'm submitting this preliminary PR to limit the review burden.

Do you mind taking a look?

Thanks,
-John

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java:
```diff
@@ -119,15 +121,43 @@ public final void removeAllTaskLevelSensors(final String taskName) {
         }
     }
 
+    public Sensor nodeLevelSensor(final String taskName, final String processorNodeName, final String sensorName, final Sensor.RecordingLevel recordingLevel, final Sensor... parents) {
```
nit: break line (too long)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java:
```diff
@@ -119,15 +121,43 @@ public final void removeAllTaskLevelSensors(final String taskName) {
+    }
+
+    public final void removeAllNodeLevelSensors(final String taskName, final String processorNodeName) {
+        final String key = threadName + "." + taskName + ".node." + processorNodeName;
```
this is the same as above: we should extract to a method to avoid inconsistencies if one might get updated but the other one slips...

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java:
```diff
@@ -152,6 +182,41 @@ public final void removeAllCacheLevelSensors(final String taskName, final String
+                                         final String sensorName,
+                                         final Sensor.RecordingLevel recordingLevel,
+                                         final Sensor... parents) {
+        final String key = threadName + "." + taskName + ".store." + storeName;
```
again :)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java:
```diff
@@ -332,6 +373,27 @@ private void addLatencyMetrics(final String scopeName, final Sensor sensor, fina
         );
     }
 
+    public static void addInvocationRateAndCount(final Sensor sensor, final String group, final Map<String, String> tags, final String operation) {
```
nit: line too long

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java:
```diff
@@ -274,31 +317,15 @@ public Sensor addThroughputSensor(final String scopeName,
 
         // first add the global operation metrics if not yet, with the global tags only
-        final Sensor parent = metrics.sensor(sensorName(buildUniqueSensorName(operationName, taskName), null), recordingLevel);
+        final Sensor parent = metrics.sensor(threadName + "." + operationName, recordingLevel);
```
Maybe I am missing something, but does this generate the same name as before? Where does `taskName` come from now?

vvcephei replied:
`taskName` was always null, and was therefore never affecting the name of the sensor.

mjsax reviewed (commented):
Some initial comments... need to wrap my head around this again. Metrics code is always confusing to me.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java:
```diff
@@ -180,70 +159,90 @@ Sensor sourceNodeForwardSensor() {
+                                                                           final String group,
+                                                                           final String taskName,
+                                                                           final String processorNodeName,
+                                                                           final Map<String, String> taskTags,
```
The last two parameters name are confusing: should they just be `allTagMap` and `tagMap`?

guozhangwang replied:
Note that although processor-level metrics parent with "task=[taskName], processor-node-id=all" is actually the same for task-level `task=[taskName]`, today we still record them as two sensors (I have a proposal for cleaning up them in the future).

vvcephei replied:
I chose those names to match the method name: `createTaskAndNodeLatencyAndThroughputSensors`. Since we are creating "task" and "node" sensors, we pass in "task" and "node tags.

I felt that `allTagMap` and `tagMap` weren't self-expanatory enough in this interface.

Note that previously, we constructed maps with those names in a more generic context, where we didn't know, for example, that the child sensor was actually a "node" sensor, so the more generic names made more sense in that context.

If you're proposing to remove the `processor-node-id=all` tag for the rollup metric in the future, I would support that.

guozhangwang replied:
Yes I do propose it, but it would require a KIP, and I'd want to keep this PR small :)

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java:
```diff
@@ -180,70 +159,90 @@ Sensor sourceNodeForwardSensor() {
-                processorNodeName,
+            final String group = "stream-processor-node-metrics";
+            final String taskName = context.taskId().toString();
+            final Map<String, String> tagMap = metrics.tagMap("task-id", context.taskId().toString(), "processor-node" + "-id", processorNodeName);
```
Why not just `"processor-node-id"`?

vvcephei replied:
oh, haha. yes...

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java:
```diff
@@ -230,38 +288,23 @@ public Sensor addLatencyAndThroughputSensor(final String scopeName,
 
         // add the operation metrics with additional tags
-        final Sensor sensor = metrics.sensor(sensorName(buildUniqueSensorName(operationName, taskName), entityName), recordingLevel, parent);
+        final Sensor sensor = metrics.sensor(entityName + "-" + (threadName + "." + operationName), recordingLevel, parent);
```
I'm +1 to @mjsax 's comment: could we centralize all the logic of constructing the sensorName, metricsName, group, and the keyName of our own bookkeeping maps, so that they are not scattered and in danger of introducing bugs?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java:
```diff
@@ -308,9 +335,26 @@ public Sensor addThroughputSensor(final String taskName,
-    private String buildUniqueSensorName(String operationName, String taskName) {
-        String task = taskName == null ? "" : taskName + ".";
-        return threadName + "." + task + operationName;
+    public static void addAvgMaxLatency(final Sensor sensor,
```
The `addAvgMaxLatency` and `addLatencyMetrics` seems duplicated logic now?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredSessionStore.java:
```diff
@@ -65,26 +69,28 @@ public void init(final ProcessorContext context, final StateStore root) {
         } finally {
-            this.metrics.recordLatency(restoreTime, startNs, time.nanoseconds());
+            this.metrics.recordLatency(
+                createTaskAndStoreLatencyAndThroughputSensors(DEBUG, "restore", metrics, metricsGroup, taskName, name(), taskTags, storeTags),
```
Why create this sensor on-the-fly instead of like others?

guozhangwang replied:
Same below.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java:
```diff
@@ -180,70 +159,90 @@ Sensor sourceNodeForwardSensor() {
-                context.taskId().toString(),
-                "processor-node",
+
+            // note: this metric can be removed in the future, as it is only recorded before being immediately removed
```
I've added the destruction sensor back with this note to future-us.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java:
```diff
@@ -46,7 +46,12 @@
+    private final Map<String, Deque<String>> storeLevelSensors = new HashMap<>();
+
+    private static final String SENSOR_PREFIX_DELIMITER = ".";
+    private static final String SENSOR_NAME_DELIMITER = ".s.";
```
I've made the following attempts to codify the sensor names:
* moved the delimiters here
* created a method for each level's prefix.
* (re)created methods for the external parent and child sensor names
* namespaced the internal sensors "internal" and the external sensors "external"

vvcephei commented:
@mjsax @guozhangwang I believe I've addressed all the outstanding concerns. I've also rebased this PR.

vvcephei committed:
finalizing

vvcephei commented:
@guozhangwang 

In that last commit:
* I simplified the handling of tags in the `StreamsMetricsImpl`
* I checked each "level" to make sure that all the sensors that are created get removed at the relevant `close()`
* During that survey, I noticed that the `taskClosedSensor` was never getting recorded. I think I messed it up last time I was modifying the metrics. I've added the `record` call in the relevant place.

I think this pretty much ties up this PR.

vvcephei commented:
Unrelated Java 8 failure:
```
java.nio.file.DirectoryNotEmptyException: /tmp/kafka-3971895545621395470/test-application/0_1/windowed-store/windowed-store.180000
	at sun.nio.fs.UnixFileSystemProvider.implDelete(UnixFileSystemProvider.java:242)
	at sun.nio.fs.AbstractFileSystemProvider.delete(AbstractFileSystemProvider.java:103)
	at java.nio.file.Files.delete(Files.java:1126)
	at org.apache.kafka.common.utils.Utils$2.postVisitDirectory(Utils.java:739)
	at org.apache.kafka.common.utils.Utils$2.postVisitDirectory(Utils.java:722)
	at java.nio.file.Files.walkFileTree(Files.java:2688)
	at java.nio.file.Files.walkFileTree(Files.java:2742)
	at org.apache.kafka.common.utils.Utils.delete(Utils.java:722)
	at org.apache.kafka.streams.processor.internals.StandbyTaskTest.cleanup(StandbyTaskTest.java:167)
```

But that is a Streams test, so we might want to track it as a flaky test.

mjsax commented:
@vvcephei This need a rebase. Thx.

vvcephei commented:
@mjsax rebased. Thanks!

guozhangwang commented:
Triggered benchmark test for this branch as it removed the delegate function which is on the critical path:

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1894

guozhangwang commented:
Perf numbers are comparable to trunk, merging to trunk now.
